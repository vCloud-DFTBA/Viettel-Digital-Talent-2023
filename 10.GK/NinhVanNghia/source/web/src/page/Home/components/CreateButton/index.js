import * as Yup from 'yup';
import { useState } from "react";
import { useFormik } from 'formik';
import { toast } from "react-toastify";
import { Button } from "react-bootstrap";

import axiosClient from 'api/axiosClient';
import AppModal from 'components/AppModal';
import { Input } from "components/FormikField/Input";
import queryClient from 'api/queryClient';
import { attendeeKeys } from 'api/attendeeQuery';


function CreateButton({className}) {
    const [show, setShow] = useState(false);
    const formik = useFormik({
        initialValues: {
            username: '',
            fullname: '',
            birthYear: 2000,
            university: '',
            major: '',
        },
        validationSchema: Yup.object({
            username: Yup.string().required('Bắt buộc').matches(/^[A-Za-z0-9]+$/, 'username chỉ chứa chữ, chữ hoa và số'),
            fullname: Yup.string().required('Bắt buộc').trim(),
            birthYear: Yup.number().min(1900, 'Năm sinh không hợp lệ').max(2100, 'Năm sinh không hợp lệ'),
            university: Yup.string().optional().trim(),
            major: Yup.string().optional().trim(),
        }),
        async onSubmit(values) {
            await axiosClient.post(`/attendee`, values);
            await queryClient.invalidateQueries(attendeeKeys.list())
            toast.success('Thêm mới thành công')
        }
    })
    return (
        <>
            <Button className={`${className} font-weight-bold`} variant="primary"
                onClick={() => setShow(true)}
            >
                Thêm người tham dự
            </Button>
            <AppModal title="Thêm người tham dự" show={show} setShow={setShow} formik={formik}>
                <div className='d-flex flex-column'>
                    <Input id="username" name="username" placeholder="Username" formik={formik}/>
                    <Input id="fullname" name="fullname" placeholder="Họ và tên" formik={formik}/>
                    <Input id="birthYear" name="birthYear" type="number" placeholder="Năm sinh" formik={formik}/>
                    <Input id="university" name="university" placeholder="Tên trường" formik={formik}/>
                    <Input id="major" name="major" placeholder="Chuyên ngành" formik={formik}/>
                </div>
            </AppModal>
        </>
    )
}


export default CreateButton
