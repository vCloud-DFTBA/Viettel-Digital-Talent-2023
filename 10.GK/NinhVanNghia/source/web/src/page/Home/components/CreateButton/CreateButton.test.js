import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';

import CreateButton from '.';


test("It should render", async () => {
    expect(renderer.create(<CreateButton/>).toJSON()).toMatchSnapshot()
})


test("It should open form", async () => {
    render(<CreateButton/>)

    // Click open form
    await userEvent.click(screen.getByText("Thêm người tham dự"))

    // Wait and check
    await screen.findByTestId('app-modal')
})
