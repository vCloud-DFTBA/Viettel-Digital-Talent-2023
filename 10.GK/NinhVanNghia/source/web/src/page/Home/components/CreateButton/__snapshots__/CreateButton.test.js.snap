// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It should render 1`] = `
<button
  className="undefined font-weight-bold btn btn-primary"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Thêm người tham dự
</button>
`;
